### PR TITLE
add --volume support for hyperctl

### LINKED ISSUE #341: add volume support in `hyperctl run` command line

### PR DESCRIPTION
**Pull request by crook**
use bind mount to close #341 

Now support: hyperctl run -v /host-src:/container-dest -t <images>

Test log:
http://pastebin.com/Wr1mZwEa

### CODE REVIEW

File: client/run.go [hunk 1]
```diff
@@ -290,6 +315,50 @@ func (cli *HyperClient) JsonFromCmdline(cmdArgs, cmdEnvs, cmdPortmaps []string,
...
+		// cmd: -v host-src:container-dest
+		srcName = fields[0]
+		destPath = fields[1]
+		readOnly = "rw"
+	} else if len(fields) == 1 {
+		// -v container-dest
+		destPath = fields[0]
+		readOnly = "rw"
+		_, filename = path.Split(destPath)
+		srcName = path.Join("/var/tmp/hyper", filename+utils.RandStr(5, "number"))
```
Comment (laijs) [inline client/run.go]: please set the srcName to be "" in this case

File: client/run.go [hunk 2]
```diff
@@ -254,6 +258,27 @@ func (cli *HyperClient) JsonFromCmdline(cmdArgs, cmdEnvs, cmdPortmaps []string,
...
+	for _, v := range cmdVols {
+		volRef, err := parseVolumes(v)
+		if err != nil {
+			return "", err
+		}
+		// Bind mount's driver is 'vfs' so far
+		var vol = pod.UserVolume{}
+		vol.Name = volRef.Volume
+		vol.Source = volRef.Volume
+		vol.Driver = "vfs"
```
Comment (laijs) [inline client/run.go]: when vol.Source is "", vol.Driver should be "", the hyperd will use its default volume driver

File: client/run.go [hunk 3]
```diff
@@ -290,6 +315,50 @@ func (cli *HyperClient) JsonFromCmdline(cmdArgs, cmdEnvs, cmdPortmaps []string,
 	return string(jsonString), nil
 }
 
+func parseVolumes(volStr string) (*pod.UserVolumeReference, error) {
```
Comment (laijs) [inline client/run.go]: it seams better that it also returns the pod.UserVolume.

File: client/run.go [hunk 4]
```diff
@@ -290,6 +315,50 @@ func (cli *HyperClient) JsonFromCmdline(cmdArgs, cmdEnvs, cmdPortmaps []string,
...
+		readOnly string
+		filename string
+	)
+
+	fields := strings.Split(volStr, ":")
+	if len(fields) == 3 {
+		// cmd: -v host-src:container-dest:rw
+		srcName = fields[0]
+		destPath = fields[1]
+		readOnly = fields[2]
```
Comment (laijs) [inline client/run.go]: please check the readOnly here.

**Comment by crook** [comment]
@laijs @gnawux please help check the new version. Thanks

File: client/run.go [hunk 5]
```diff
@@ -290,6 +303,66 @@ func (cli *HyperClient) JsonFromCmdline(cmdArgs, cmdEnvs, cmdPortmaps []string,
...
+		// -v container-dest
+		srcName = ""
+		destPath = fields[0]
+		readOnly = "rw"
+	} else {
+		return nil, nil, fmt.Errorf("flag format should be like : --volume=[host-src:]container-dest[:rw|ro]")
+	}
+
+	vol.Source = srcName
+	_, vol.Name = path.Split(vol.Source)
```
Comment (laijs) [inline client/run.go]: it might need to be a name with order as suffix or a random name

File: client/run.go [hunk 6]
```diff
@@ -290,6 +303,66 @@ func (cli *HyperClient) JsonFromCmdline(cmdArgs, cmdEnvs, cmdPortmaps []string,
...
+			return nil, nil, fmt.Errorf("flag only support(ro or rw): --volume")
+		}
+	} else if len(fields) == 2 {
+		// cmd: -v host-src:container-dest
+		srcName = fields[0]
+		destPath = fields[1]
+		readOnly = "rw"
+	} else if len(fields) == 1 {
+		// -v container-dest
+		srcName = ""
```
Comment (laijs) [inline client/run.go]: in golang, "" is the initial/default value of a string. please remove this line.

File: client/run.go [hunk 7]
```diff
@@ -290,6 +303,66 @@ func (cli *HyperClient) JsonFromCmdline(cmdArgs, cmdEnvs, cmdPortmaps []string,
...
 }
 
+func parseVolume(volStr string) (*pod.UserVolume, *pod.UserVolumeReference, error) {
+
+	var (
+		vol      = pod.UserVolume{}
+		volRef   = pod.UserVolumeReference{}
+		srcName  string
+		destPath string
+		readOnly string
```
Comment (laijs) [inline client/run.go]: readOnly = false.
and set it to true in the first branch below when needed.

File: client/run.go [hunk 8]
```diff
@@ -290,6 +303,66 @@ func (cli *HyperClient) JsonFromCmdline(cmdArgs, cmdEnvs, cmdPortmaps []string,
 	return string(jsonString), nil
 }
 
+func parseVolume(volStr string) (*pod.UserVolume, *pod.UserVolumeReference, error) {
+
+	var (
+		vol      = pod.UserVolume{}
+		volRef   = pod.UserVolumeReference{}
```
Comment (laijs) [inline client/run.go]: add volDriver string

File: client/run.go [hunk 9]
```diff
@@ -290,6 +303,66 @@ func (cli *HyperClient) JsonFromCmdline(cmdArgs, cmdEnvs, cmdPortmaps []string,
...
+		destPath = fields[0]
+		readOnly = "rw"
+	} else {
+		return nil, nil, fmt.Errorf("flag format should be like : --volume=[host-src:]container-dest[:rw|ro]")
+	}
+
+	vol.Source = srcName
+	_, vol.Name = path.Split(vol.Source)
+
+	vol.Driver = "vfs"
```
Comment (laijs) [inline client/run.go]:
```
vol := pod.UserVolume{
        Driver : volDriver,
        Source: srcName,
      ....
}
```

volRef needs to be like this also.

**Comment by crook** [comment]
@laijs Thanks for your comments. and I have refined the codes. please help check

File: client/run.go [hunk 10]
```diff
@@ -306,58 +306,59 @@ func (cli *HyperClient) JsonFromCmdline(cmdArgs, cmdEnvs, cmdPortmaps []string,
...
-	_, vol.Name = path.Split(vol.Source)
-
-	vol.Driver = "vfs"
-	if vol.Source == "" {
+	_, volName = path.Split(srcName)
+	if srcName == "" {
 		// Set default volume driver and use destPath as volume Name
-		vol.Driver = ""
-		_, vol.Name = path.Split(destPath)
+		volDriver = ""
```
Comment (laijs) [inline client/run.go]: move `volDriver = ""` up in one of the branches.

Comment (crook) [inline client/run.go]: what if '-v "":/hello' format? the driver is empty or VFS?

File: client/run.go [hunk 11]
```diff
@@ -254,6 +258,15 @@ func (cli *HyperClient) JsonFromCmdline(cmdArgs, cmdEnvs, cmdPortmaps []string,
 		}
 	}
 
+	for _, v := range cmdVols {
+		vol, volRef, err := parseVolume(v)
+		if err != nil {
+			return "", err
+		}
+		volumes = append(volumes, *vol)
+		volumesRef = append(volumesRef, *volRef)
```
Comment (laijs) [inline client/run.go]: I prefer to generate the volume name here. 
fmt.Sprintf("vol-%d", i); `i` is the order of the volumes.

File: client/run.go [hunk 12]
```diff
@@ -290,6 +303,67 @@ func (cli *HyperClient) JsonFromCmdline(cmdArgs, cmdEnvs, cmdPortmaps []string,
 	return string(jsonString), nil
 }
 
+func parseVolume(volStr string) (*pod.UserVolume, *pod.UserVolumeReference, error) {
```
Comment (laijs) [inline client/run.go]: just pod.UserVolume is Ok, don't need to be pointer

Comment (crook) [inline client/run.go]: The parseVolume is like parsePortMapping, I prefer to keep consistent. And pointer is faster and less resouce

File: client/run.go (hunk 11 above)
Comment (laijs) [inline client/run.go]: the order doesn't matter, but it is the source for the uniq identification.

File: client/run.go (hunk 12 above)
Comment (laijs) [inline client/run.go]: speed and resouce are unimportant here.
And in golang, pointer for small object is slower and uses more resource, golang always allocate heap memory for pointers.

File: client/run.go (hunk 11 above)
Comment (crook) [inline client/run.go]: make sense. But it's better to put all the codes into one function

File: client/run.go (hunk 10 above)
Comment (laijs) [inline client/run.go]: Good catch! it is empty. ignore my request.

File: client/run.go (hunk 11 above)
Comment (laijs) [inline client/run.go]: just add a volName argument to the parseVolumes()

Comment (crook) [inline client/run.go]: There is already one volStr argument. User will be confused if add more volName, right?
. The vol Name will be aded 5 random number in routine parseVolume, which is enough to make the name uniq.

File: client/run.go (hunk 12 above)
Comment (crook) [inline client/run.go]: make sense, do you need me also update the parsePortMapping routine? if no need. please help close the PR, thanks.

File: client/run.go [hunk 13]
```diff
@@ -5,6 +5,7 @@ import (
 	"fmt"
 	"io/ioutil"
 	"os"
+	"path"
```
Comment (laijs) [inline client/run.go]: use path/filepath

File: client/run.go [hunk 14]
```diff
@@ -290,6 +303,67 @@ func (cli *HyperClient) JsonFromCmdline(cmdArgs, cmdEnvs, cmdPortmaps []string,
...
+		srcName = fields[0]
+		destPath = fields[1]
+	} else if len(fields) == 1 {
+		// -v container-dest
+		destPath = fields[0]
+	} else {
+		return nil, nil, fmt.Errorf("flag format should be like : --volume=[host-src:]container-dest[:rw|ro]")
+	}
+
+	_, volName = path.Split(srcName)
```
Comment (laijs) [inline client/run.go]: how about moving it to else branch ?

Comment (crook) [inline client/run.go]: can't catch you

File: client/run.go (hunk 12 above)
Comment (crook) [inline client/run.go]: won't update this. got this puzzle if I don't use pointer (): https://www.reddit.com/r/golang/comments/2xmnvs/returning_nil_for_a_struct/

File: client/run.go (hunk 13 above)
Comment (crook) [inline client/run.go]: you are right. I should use path/filepath.Join

**Comment by laijs** [comment]
could you squash the commits into one please.

File: client/run.go (hunk 14 above)
Comment (laijs) [inline client/run.go]: here are two "_, volName =", the value depends on the "if else" branches, could you move it into the branch?

Comment (crook) [inline client/run.go]: got you point. I will update

**Comment by crook** [comment]
I would like to git rebase. but don't know what's the fast/correct way? Can you advise?

File: client/run.go [hunk 15]
```diff
@@ -290,6 +303,68 @@ func (cli *HyperClient) JsonFromCmdline(cmdArgs, cmdEnvs, cmdPortmaps []string,
...
+		// Set default volume driver and use destPath as volume Name
+		volDriver = ""
+		_, volName = filepath.Split(destPath)
+	} else {
+		_, volName = filepath.Split(srcName)
+		// Auto create the source folder on the host , otherwise hyperd will complain
+		if _, err := os.Stat(srcName); err != nil && os.IsNotExist(err) {
+			if err := os.MkdirAll(srcName, os.FileMode(0777)); err != nil {
+				return nil, nil, err
+			}
```
Comment (laijs) [inline client/run.go]: please convert the source path to abspath

Comment (laijs) [inline client/run.go]: and please check the destPath, ensure it starts with /

Comment (crook) [inline client/run.go]: ok. #431

**Pull request merged**